File: frontend/src/infrastructure/repositories/location_repository.ts
/**
 * Repositorio de Ubicaciones.
 * Gestiona las operaciones CRUD de ubicaciones con el backend.
 */

import api_client from '../api_client';
import type {
    Location_Model,
    Location_List_Response
} from '../../domain/models';

/**
 * Obtiene todas las ubicaciones.
 * @param owner_email - Email del propietario para filtrar (opcional).
 * @returns Promesa con la lista de ubicaciones.
 */
export const get_all_locations = async (owner_email?: string): Promise<Location_List_Response> => {
    const params = owner_email ? { owner_email } : {};
    const response = await api_client.get<Location_List_Response>('/locations', { params });
    return response.data;
};

/**
 * Obtiene las ubicaciones del usuario autenticado.
 * @returns Promesa con la lista de ubicaciones propias.
 */
export const get_my_locations = async (): Promise<Location_List_Response> => {
    const response = await api_client.get<Location_List_Response>('/locations/my');
    return response.data;
};

/**
 * Obtiene una ubicación por su ID.
 * @param location_id - ID de la ubicación.
 * @returns Promesa con los detalles de la ubicación.
 */
export const get_location_by_id = async (location_id: string): Promise<Location_Model> => {
    const response = await api_client.get<Location_Model>(`/locations/${location_id}`);
    return response.data;
};

/**
 * Crea una nueva ubicación con imagen opcional.
 * @param title - Título del marcador.
 * @param address - Dirección a geocodificar.
 * @param description - Descripción opcional.
 * @param image - Archivo de imagen opcional.
 * @returns Promesa con la ubicación creada.
 */
export const create_location = async (
    title: string,
    address: string,
    description?: string,
    image?: File
): Promise<Location_Model> => {
    const form_data = new FormData();
    form_data.append('title', title);
    form_data.append('address', address);

    if (description) {
        form_data.append('description', description);
    }

    if (image) {
        form_data.append('image', image);
    }

    const response = await api_client.post<Location_Model>('/locations', form_data, {
        headers: {
            'Content-Type': 'multipart/form-data',
        },
    });

    return response.data;
};

/**
 * Actualiza una ubicación existente.
 * @param location_id - ID de la ubicación a actualizar.
 * @param data - Datos a actualizar.
 * @returns Promesa con la ubicación actualizada.
 */
export const update_location = async (
    location_id: string,
    data: { title?: string; description?: string; address?: string }
): Promise<Location_Model> => {
    const response = await api_client.put<Location_Model>(`/locations/${location_id}`, data);
    return response.data;
};

/**
 * Elimina una ubicación.
 * @param location_id - ID de la ubicación a eliminar.
 * @returns Promesa que se resuelve cuando se elimina.
 */
export const delete_location = async (location_id: string): Promise<void> => {
    await api_client.delete(`/locations/${location_id}`);
};

File: backend/repositories/location_repository.py
"""
Repositorio para operaciones de Ubicación en MongoDB.
"""

from datetime import datetime
from bson import ObjectId
from motor.motor_asyncio import AsyncIOMotorDatabase
from models.location import Location


class LocationRepository:
    """
    Repositorio para gestionar ubicaciones en la base de datos.
    """
    
    def __init__(self, db: AsyncIOMotorDatabase):
        """
        Inicializa el repositorio con la conexión a la base de datos.
        :param db: Instancia de la base de datos MongoDB.
        """
        self.collection = db["locations"]
    
    async def find_all(self, owner_email: str | None = None) -> list[Location]:
        """
        Obtiene todas las ubicaciones, opcionalmente filtradas por propietario.
        :param owner_email: Email del propietario para filtrar (opcional).
        :return: Lista de ubicaciones.
        """
        query = {}
        if owner_email:
            query["owner_email"] = owner_email
        
        cursor = self.collection.find(query).sort("created_at", -1)
        locations = []
        async for doc in cursor:
            doc["_id"] = str(doc["_id"])
            locations.append(Location(**doc))
        return locations
    
    async def find_by_id(self, location_id: str) -> Location | None:
        """
        Busca una ubicación por su ID.
        :param location_id: ID de la ubicación a buscar.
        :return: Ubicación encontrada o None.
        """
        try:
            doc = await self.collection.find_one({"_id": ObjectId(location_id)})
            if doc:
                doc["_id"] = str(doc["_id"])
                return Location(**doc)
        except Exception:
            pass
        return None
    
    async def create(self, location: Location) -> Location:
        """
        Crea una nueva ubicación.
        :param location: Datos de la ubicación a crear.
        :return: Ubicación creada con su ID.
        """
        location_dict = location.model_dump(exclude={"id"})
        location_dict["created_at"] = datetime.utcnow()
        location_dict["updated_at"] = datetime.utcnow()
        
        result = await self.collection.insert_one(location_dict)
        location_dict["_id"] = str(result.inserted_id)
        return Location(**location_dict)
    
    async def update(self, location_id: str, update_data: dict) -> Location | None:
        """
        Actualiza una ubicación existente.
        :param location_id: ID de la ubicación a actualizar.
        :param update_data: Datos a actualizar.
        :return: Ubicación actualizada o None.
        """
        update_data["updated_at"] = datetime.utcnow()
        
        try:
            await self.collection.update_one(
                {"_id": ObjectId(location_id)},
                {"$set": update_data}
            )
            return await self.find_by_id(location_id)
        except Exception:
            return None
    
    async def delete(self, location_id: str) -> bool:
        """
        Elimina una ubicación.
        :param location_id: ID de la ubicación a eliminar.
        :return: True si se eliminó, False si no existía.
        """
        try:
            result = await self.collection.delete_one({"_id": ObjectId(location_id)})
            return result.deleted_count > 0
        except Exception:
            return False
